[#125960675] Array Constraint Derivation (Array Pt 1)

cramja opened the pull request:
### Background
ORCA propagates filters across joins when the filter is applied to a join key. In essence, 
```SQL
-- (1)
SELECT * FROM foo, bar WHER foo.id = bar.id AND bar.id IN (1, 20);
```
is equivalent to
```SQL
-- (2)
SELECT * FROM foo, bar WHER foo.id = bar.id AND bar.id IN (1, 20) AND foo.id IN (1,20);
```
however ORCA **currently expands arrays** meaning that (1) actually is transformed to
```SQL
-- (3)
SELECT * FROM foo, bar WHER foo.id = bar.id AND bar.id IN (1, 20) AND foo.id = 1 OR foo.id = 20;
```

For more background, check out [this document](https://docs.google.com/document/d/1l61eaud_qtC15TZLcAf17dwCZ6hCU1zdS3Q4GfXC_2k/edit#).

### Change
Previously, ORCA propagated array IN constraints (x in (1,2,3)) by first expanding them into a disjunction of equality statements (x = 1 OR x = 2 ...). This process was slow, especially with large IN lists. This change propagates the array expression. As in, now we derive an array expression, instead of the more heavyweight OR statement.

### Process

This pull request comes in 3 commits. The first commit are helper methods. They are used by the next two commits.

The second commit introduces the main functionality in CConstraintInterval. The four methods to look at are:
- PexprConstructArrayScalar which converts a constraint to an expression
- PcnstrIntervalFromScalarArrayCmp which converts a scalar array comparison to a constraint interval
- convertsToIn/convertsToNotIn which are helper methods for determining if the ConstraintInterval can convert to an Array comparison expression (as opposed to at disjunction of greater/less than functions

The third commit introduces tests for this functionality.

A traceflag has been added to enable this functionality- disabled by default. Later pull requests will fully integrate the array derivation feature.

cramja commented:
@vraghavan78 @lpetrov-pivotal @hsyuan @d @oarap 

we can hold a code review session if needed for context on this PR

hsyuan commented:
Good job! We can have a code review session.

vraghavan78 commented on data/dxl/minidump/IN-ArrayCmp.mdp:
```diff
@@ -311,30 +311,14 @@
+              <dxl:ConstValue TypeMdid="0.23.1.0" IsNull="false" IsByValue="true" Value="2"/>
+              <dxl:ConstValue TypeMdid="0.23.1.0" IsNull="false" IsByValue="true" Value="3"/>
+            </dxl:Array>
+          </dxl:ArrayComp>
```
nice!

vraghavan78 commented on data/dxl/minidump/IN-ArrayCmp.mdp:
```diff
@@ -363,20 +347,14 @@
-            </dxl:Comparison>
-          </dxl:Or>
+            </dxl:Array>
+          </dxl:ArrayComp>
```
sweet!

vraghavan78 commented on libgpopt/include/gpopt/base/CConstraintInterval.h:
```diff
@@ -84,6 +92,8 @@ namespace gpopt
 			virtual
 			CExpression *PexprConstructScalar(IMemoryPool *pmp) const;
 
+			CExpression *PexprConstructArrayScalar(IMemoryPool *pmp) const;
```
virtual?

vraghavan78 commented on libgpopt/include/gpopt/base/CConstraintInterval.h:
```diff
@@ -128,6 +138,15 @@ namespace gpopt
 									CExpression *pexpr,
 									CColRef *pcr
 									);
+
```
can you add a comment?

vraghavan78 commented on libgpopt/include/gpopt/base/CConstraintInterval.h:
```diff
@@ -216,6 +235,12 @@ namespace gpopt
 			CConstraint *PcnstrRemapForColumn(IMemoryPool *pmp, CColRef *pcr) const;
 
+			// converts to an array in expression
+			bool convertsToNotIn() const;
```
FConvertsToNotIn

vraghavan78 commented on libgpopt/include/gpopt/base/CConstraintInterval.h:
```diff
@@ -216,6 +235,12 @@ namespace gpopt
+			bool convertsToNotIn() const;
+
+			// converts to an array not in expression
+			bool convertsToIn() const;
```
FConvertsToIn

vraghavan78 commented on libgpopt/include/gpopt/base/CConstraintInterval.h:
```diff
@@ -257,7 +282,31 @@ namespace gpopt
 				CColRef *pcr = NULL
 				);
 
+			//  ConstraintInterval from the given expression
```
generate constraint interval from ...

cramja commented on libgpopt/include/gpopt/base/CUtils.h:
```diff
@@ -748,6 +748,9 @@ namespace gpopt
 			BOOL FComparisonPossible(DrgPcr *pdrgpcr, IMDType::ECmpType ecmpt);
 
+			static
+			ULONG FCountOperator(CExpression *pexpr, INT Eopid);
```
UlCountOperator or UlOperatorCount

cramja commented on libgpopt/src/base/CRange.cpp:
```diff
@@ -241,6 +241,38 @@ CRange::FOverlapsRight
+//		CRange::FRightEqualsLeft
+//
+//	@doc:
+//		Checks if the right element equal the left element of the given range.
```
equals

cramja commented on libgpopt/src/base/CRange.cpp:
```diff
@@ -241,6 +241,38 @@ CRange::FOverlapsRight
+//
+//	@doc:
+//		Checks if the right element equal the left element of the given range.
+//      This is useful when checking if 2 ranges touch at the ends
```
give an example

(-inf, 8)(8, inf)

rename to FUpperBoundEqualsLowerBound or something similar (right/left not good)

Add comment to say if the boundries are inclusive or exclusive.

cramja commented on libgpopt/src/base/CUtils.cpp:
```diff
@@ -4094,6 +4094,35 @@ CUtils::FComparisonPossible
+		INT Eopid
+	)
+{
+	INT iopCnt = 0;
```
Hungarian iopCnt -> iOpCnt

cramja commented on libgpopt/src/base/CUtils.cpp:
```diff
@@ -4094,6 +4094,35 @@ CUtils::FComparisonPossible
+	}
+
+
+	for (ULONG ulChild = 0; ulChild < pexpr->UlArity(); ulChild++)
```
pull pexpr->UlArity() out of for loop

cramja commented on libgpopt/src/base/CConstraint.cpp:
```diff
@@ -185,14 +184,14 @@ CConstraint::PcnstrFromScalarExpr
+
+		// try creating a single constraint from the expression
+		pcnstr = CConstraintInterval::PciIntervalFromScalarExpr(pmp, pexpr, pcr);
+		if (NULL == pcnstr && CUtils::FScalarArrayCmp(pexpr))
```
redundant, remove these four lines

cramja replied:
just call PciIntervalFromScalarExpr

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -136,30 +139,158 @@ CConstraintInterval::PciIntervalFromScalarExpr
 	GPOS_ASSERT(1 >= pdpScalar->PcrsUsed()->CElements());
 #endif //GPOS_DEBUG
-
+	CConstraintInterval *pci;
```
init to NULL

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -136,30 +139,158 @@ CConstraintInterval::PciIntervalFromScalarExpr
+			{
+				pci = CConstraintInterval::PcnstrIntervalFromScalarArrayCmp(pmp, pexpr, pcr);
+			}
+			else
```
remove when already init'd to NULL

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -136,30 +139,158 @@ CConstraintInterval::PciIntervalFromScalarExpr
+	CScalarArrayCmp::EArrCmpType earrccmpt = popScArrayCmp->Earrcmpt();
+
+	if (!CPredicateUtils::FCompareIdentToConstArray(pexpr) ||
+		!(CScalarArrayCmp::EarrcmpAny == earrccmpt  || CScalarArrayCmp::EarrcmpAll == earrccmpt))
```
comment the code: We do not handle IDF (is distinct from)

cramja replied:
file story for IDF into ConstraintInterval

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -136,30 +139,158 @@ CConstraintInterval::PciIntervalFromScalarExpr
+	// construct ranges representing IN or NOT IN
+	DrgPrng *prgrng = GPOS_NEW(pmp) DrgPrng(pmp);
+	const IComparator *pcomp = COptCtxt::PoctxtFromTLS()->Pcomp();
+
```
Add a switch, blow up on default case.

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -136,30 +139,158 @@ CConstraintInterval::PciIntervalFromScalarExpr
+	}
+	else
+	{
+		// IN case, create ranges [X, X] [Y, Y] [Z, Z]
```
logically, equality would make sense to come first in the Switch

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -136,30 +139,158 @@ CConstraintInterval::PciIntervalFromScalarExpr
+			}
+
+			pdatum = (*rngdatum)[ul];
+			pdatum->AddRef();
```
add duplicate check?

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -257,33 +388,13 @@ CConstraintInterval::PciIntervalFromColConstCmp
-
-	// comparison operator
-	if (IMDType::EcmptOther == ecmpt)
+	DrgPrng *pdrngprng = PciRangeFromColConstCmp(pmp, ecmpt, popScConst);
```
CConstraintInterval *pcri = NULL;

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -257,33 +388,13 @@ CConstraintInterval::PciIntervalFromColConstCmp
-	{
-		pdatum->AddRef();
-		pdrgprng->Append(GPOS_NEW(pmp) CRange(pcomp, ecmpt, pdatum));
+		return GPOS_NEW(pmp) CConstraintInterval(pmp, pcr, pdrngprng, false /*fIncludesNull*/);
```
pcri =  ...

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -257,33 +388,13 @@ CConstraintInterval::PciIntervalFromColConstCmp
-		pdrgprng->Append(GPOS_NEW(pmp) CRange(pcomp, ecmpt, pdatum));
+		return GPOS_NEW(pmp) CConstraintInterval(pmp, pcr, pdrngprng, false /*fIncludesNull*/);
 	}
+	return NULL;
```
return pcri;

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -508,6 +620,18 @@ CConstraintInterval::PexprConstructScalar
 		return CUtils::PexprScalarConstBool(pmp, false /*fval*/, false /*fNull*/);
 	}
 
+	// remove array derive to expression
```
duhhhh

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -508,6 +620,18 @@ CConstraintInterval::PexprConstructScalar
+	// remove array derive to expression
+	if (GPOS_FTRACE(EopttraceEnableArrayDerive))
+	{
+		// this is a candidate for conversion to an array constraint
```
remove this comment?

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CConstraintInterval::PexprConstructScalar (private)
```
function name different

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+CConstraintInterval::PexprConstructArrayScalar(IMemoryPool *pmp, bool isIn) const {
+	CMDAccessor *pmda = COptCtxt::PoctxtFromTLS()->Pmda();
+
+	ULONG ulRngs = m_pdrgprng->UlLength();
```
const

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+	ULONG ulRngs = m_pdrgprng->UlLength();
+	IMDType::ECmpType eCmpType;
+	CScalarArrayCmp::EArrCmpType eArrayCmpType;
+
```
init the above variable to sentinel

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+//
+//---------------------------------------------------------------------------
+CExpression *
+CConstraintInterval::PexprConstructArrayScalar(IMemoryPool *pmp, bool isIn) const {
```
fIn

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+	CMDAccessor *pmda = COptCtxt::PoctxtFromTLS()->Pmda();
+
+	ULONG ulRngs = m_pdrgprng->UlLength();
+	IMDType::ECmpType eCmpType;
```
ecmptype

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+
+	ULONG ulRngs = m_pdrgprng->UlLength();
+	IMDType::ECmpType eCmpType;
+	CScalarArrayCmp::EArrCmpType eArrayCmpType;
```
earrcmptype

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+		eCmpType = IMDType::EcmptEq;
+		eArrayCmpType = CScalarArrayCmp::EarrcmpAny;
+	}
+	else
```
if (!fIn)

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+	if (isIn)
+	{
+		eCmpType = IMDType::EcmptEq;
+		eArrayCmpType = CScalarArrayCmp::EarrcmpAny;
```
move this as the initializer

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+	// [x,x], ... ,[y,y] or the NOT IN case (-inf, x),(x,y), ... ,(z,inf).
+	for (ULONG ul = 0; ul < ulRngs; ul++)
+	{
+		IDatum *pdtm = (*m_pdrgprng)[ul]->PdatumRight();
```
pdatum

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+//      It is a mistake to call this method without first detecting if the
+//      stored ranges can be converted to an IN or NOT in statement. The param
+//		'isIn' refers to the statement being an IN statement, and if set to false,
+// 		it is considered a NOT IN statement
```
please check if the input is kosher for IN and Not In

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -542,6 +666,201 @@ CConstraintInterval::PexprConstructScalar
+										GPOS_NEW(pmp) CScalarArray(pmp, pmdidColType, pmdidArrType, false /*fMultiDimensional*/),
+										prngexpr
+										);
+
```
refactor the common stuff into a util

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -1101,4 +1420,49 @@ CConstraintInterval::OsPrint
+//	@function:
+//		CRange::PciRangeFromColConstCmp
+//
+//	@doc:
```
give an example (a = 1) OR (a = 2)

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -1101,4 +1420,49 @@ CConstraintInterval::OsPrint
+		pdatum->AddRef();
+		pdrgprng->Append(GPOS_NEW(pmp) CRange(pcomp, ecmpt, pdatum));
+	}
+
```
consider refactoring common part. fix the comment

cramja commented:
LGTM


- Jesse & Venky

cramja commented:
@vraghavan78 @d 
* I addressed hungarian and simple logic issues in the [first new commit](https://github.com/greenplum-db/gporca/pull/74/commits/17b437e4fb89cd078fd96401a91a6b9ece383da7)

* Then I made a test for the removal of duplicates we talked about and implemented removal of duplicates. So x IN (1,1,1,2) will be represented as a constraint like [[1,1],[2,2]].

* Then I refactored out the common code we saw in the creation of ArrayCmp expressions.

cramja commented on libgpopt/src/base/CConstraintInterval.cpp:
```diff
@@ -1101,4 +1393,49 @@ CConstraintInterval::OsPrint
+	(
+	IMemoryPool *pmp,
+	IMDType::ECmpType ecmpt,
+	const CScalarConst *popsccnst
```
popScConst
